### PR TITLE
Lecture 16: Building Blocks of Deep Learning

### PR DESCRIPTION
**Pull request by scottsun95**
The rendered version is at https://scottsun95.github.io/pgm-spring-2019/notes/lecture-16/.

This was done by me, @raunaqbhirangi, and @nvog

### CODE REVIEW

File: _posts/2019-03-18-lecture-16.md [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+      <figcaption>
+        <strong>Figure 1</strong>
+        Hierarchy of Features in CNNs
+      </figcaption>
+    </div>
+   </div>
+</figure>
+
+Examples of ConvNets show a trend towards ever increasing numbers of layers:
+* AlexNet, 8 layers
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md]: please add references to these papers and a reference section to the bottom of the notes

File: _posts/2019-03-18-lecture-16.md [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+<figure id="example-figure" class="l-body-outset">
+  <div class="row">
+    <div class="col two">
+      <img src="{{ 'assets/img/notes/lecture-16/rnn_grad_flow.png' | relative_url }}" />
+    </div>
+   </div>
+</figure>
+
+LSTMs are designed to solve the long-term dependency problem by creating a path with uninterrupted gradient flow during backpropagation.
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md]: need references

File: _posts/2019-03-18-lecture-16.md [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+Attention can be used in conjunction with conventional CNNs in image processing. Features extracted from the CNN are used as key vectors and attention is used to sequentialy compute the caption as tokens. It can also be used in image paragraph generation, which is the generation of a long paragraph to describe an image. This is a challenging task because it involves long-term reasoning about language and visual features. Each sentence needs to be grounded on visual features to ensure contentful descriptions. One technique for doing this was presented in the lecture and it proceeds as follow:
+* The image is first segmented into semantic regions, and each region is locally captioned with phrases
+* Attention is applied on the visual regions and the text phrases and the resulting features fed to a generator which generates sentences
+* These sentences are then fed to a sentence discriminator and a recurrent topic-transition discriminator for assessing sentence plausibility and topic coherence respectively
+* A paragraph description corpus is adopted to provide linguistic knowledge about paragraphgeneration, which depicts the true data distribution of the discriminators
+
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md:269]: references

File: _posts/2019-03-18-lecture-16.md [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+      <img src="{{ 'assets/img/notes/lecture-16/img-paragraph.png' | relative_url }}" />
+    </div>
+   </div>
+</figure>
+
+
+
+
+
+# Transformers: Multi-Headed Attention
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md:283]: reference

File: _posts/2019-03-18-lecture-16.md [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+<figure id="multiheadattneq" class="l-body-outset">
+  <div class="row">
+    <div class="col one">
+      <img src="{{ 'assets/img/notes/lecture-16/multiheadattneq.png' | relative_url }}" />
+    </div>
+   </div>
+</figure>
+
+
+## BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md:319]: reference

File: _posts/2019-03-18-lecture-16.md [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+
+In BERT (Bidirectional Encoder Representations from Transformers), Devlin et al. (2018) used the Transformer architecture to great advantage in order to obtain *contextualized* word 
+embeddings.
+Despite requiring a massive amount of compute power, the authors show that the embeddings can be used out-of-the-box to 
+perform at the state-of-the-art for eleven NLP tasks.
+
+Most past work involving word embeddings (e.g. *word2vec*) learns a vector representation of a word using a simple feedforward network to predict a 
+particular word's context.
+Thus, even if a word has multiple meanings dependent on context, it will only receive a single embedding vector. 
+On the other hand, BERT, similar to OpenAI's ELMo (Peters et al., 2018), builds a *contextual* word embedding by encoding a word  by taking into account the particular sentence context
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md]: elmo is by allen institute for ai and university of washington, not by open ai
open ai was the gpt model, please add all of these references and links to these papers
also give a 1-2 paragraph introduction to elmo, gpt and gpt 2 and outline the differences between elmo, bert, gpt, gpt 2: type of base model (lstm/bilstm/transformer), unsupervised training objective (language model or mask language model or predict next sentence etc.), supervised training objective (new model for each supervised task or just fine tune the unsupervised model or no need fine tune at all). read those 4 papers if you dont know the answers to these 4 questions and summarize your findings.

File: _posts/2019-03-18-lecture-16.md [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+        Image Credit: The Illustrated Transformer by Jay Alammar.
+      </figcaption>
+    </div>
+   </div>
+</figure>
+
+In the paper, the authors report using both Wikipedia data (2.5B words) and eBook data (800M words) for training an 
+Transformer encoder with hundreds of millions of parameters.
+While Google is able to train BERT in just 4 days on 4 TPU pods, training is impractical for academics with traditional GPU resources,
+For example, a standard 4 GPU desktop with an RTX 2080Ti would take almost 99 days to complete training!
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md:362]: add more details on their training objectives experimental results

File: _posts/2019-03-18-lecture-16.md [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+Multi-headed attention can then jointly attend to information from multiple different representations at different positions by:
+
+<figure id="multiheadattneq" class="l-body-outset">
+  <div class="row">
+    <div class="col one">
+      <img src="{{ 'assets/img/notes/lecture-16/multiheadattneq.png' | relative_url }}" />
+    </div>
+   </div>
+</figure>
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md]: more details on experimental results and findings

File: _posts/2019-03-18-lecture-16.md [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+  - name: Nikolai Vogler
+
+editors:
+  - name: Editor 1  # editor's full name
+    url: "#"  # optional URL to the editor's homepage
+
+---
+
+# Convolutional Neural Networks (CNNs)
+
```
Comment (pliang279) [inline _posts/2019-03-18-lecture-16.md:22]: some more details needed on this cnn layers, convolutional filters, pooling etc.

**Comment by pliang279** [comment]
hi everyone, i've reviewed these scribe notes, please make these changes, thanks!

**Comment by scottsun95** [comment]
@pliang279 Where can we see the changes you've suggested?

**Comment by alshedivat** [comment]
@pliang279, looks like your suggestions are not visible (probably, you didn't hit submit the review?)

**Comment by pliang279** [comment]
oops, can u see them now?

**Comment by nvog** [comment]
It's not letting me mark the suggested changes by @pliang279 as Resolved for Transformer/BERT but I did resolve the changes. Can you review again?

File: _posts/2019-03-18-lecture-16.md (hunk 3 above)
Comment (raunaqbhirangi) [reply _posts/2019-03-18-lecture-16.md:269]: Fixed.

**Pull request merged**